src/services/users.js

```diff
@@ -0,0 +1,178 @@
+        const user = await User.findOne({ _id: req.params.id }).exec();
+        const isMatchPasswords = await bcrypt.compare(
+          oldPassword,
+          user.password
+        );
+        if (!isMatchPasswords) {
+          return res.boom.badRequest('wrong old password');
+        }
+        if (isMatchPasswords && oldPassword !== newPassword) {
+          user.password = newPassword;
+          const salt = await bcrypt.genSalt(10);
+          user.password = await bcrypt.hash(user.password, salt);
+          const updatedUserPassword = User.findOneAndUpdate(
+            { _id: req.params.id },
+            user
+          );
+          return updatedUserPassword ? res.json(true) : res.boom.notFound();
+        } else {
+          res.boom.badRequest('new password cannot be the same as old');
+        }
+      }
+    } catch {
+      res.boom.notFound();
+    }
+  }
+
+  async blockUser (req, res) {
+    try {
+      const statusBlocked = await UserStatuses.findOne({
+        name: 'Blocked'
+      }).exec();
+      const statusDeleted = await UserStatuses.findOne({
+        name: 'Deleted'
+      }).exec();
+      const user = await User.findOne({ _id: req.params.id }).exec();
+      const currentStatus = user.status;
+      if (
+        statusBlocked._id.toString() !== currentStatus.toString() &&
+        statusDeleted._id.toString() !== currentStatus.toString()
+      ) {
+        const user = await User.findByIdAndUpdate(
+          req.params.id,
+          { status: statusBlocked._id },
+          {
+            new: true
+          }
+        ).exec();
+        return user ? res.json(true) : res.boom.notFound();
+      } else {
+        return res.boom.badRequest('user already blocked');
+      }
+    } catch (err) {
+      res.boom.notFound();
+    }
+  }
+
+  async deleteUser (req, res) {
+    try {
+      const statusDeleted = await UserStatuses.findOne({
+        name: 'Deleted'
+      }).exec();
+      const user = await User.findOne({ _id: req.params.id }).exec();
+      const currentStatus = user.status;
+      if (statusDeleted._id.toString() !== currentStatus.toString()) {
+        const user = await User.findByIdAndUpdate(
+          req.params.id,
+          { status: statusDeleted._id },
+          {
+            new: true
+          }
+        ).exec();
+        return user ? res.json(true) : res.boom.notFound();
+      } else {
+        return res.boom.badRequest('user already deleted');
+      }
+    } catch (err) {
+      res.boom.notFound();
+    }
+  }
+
+  async #createBaseUser (req, res, userRole) {
+    try {
+      const statusActive = await UserStatuses.findOne({
+        name: 'Active'
+      }).exec();
+      const newUser = new User(req.body);
+      newUser.status = statusActive._id;
+      const salt = await bcrypt.genSalt(10);
+      newUser.password = await bcrypt.hash(newUser.password, salt);
+      const userExist = User.findOne({ email: req.body.email });
```

Comment:
check if user exists at start. If user exists, but his status is deleted, we can create new with such an email

Reply:
fixed